hudi-common/src/main/java/org/apache/hudi/common/table/view/HoodieTableFileSystemView.java

```diff
@@ -199,7 +201,7 @@ protected boolean isPendingCompactionScheduledForFileId(HoodieFileGroupId fgId)
   protected void resetPendingCompactionOperations(Stream<Pair<String, CompactionOperation>> operations) {
     // Build fileId to Pending Compaction Instants
     this.fgIdToPendingCompaction = createFileIdToPendingCompactionMap(operations.map(entry ->
-      Pair.of(entry.getValue().getFileGroupId(), Pair.of(entry.getKey(), entry.getValue()))).collect(Collectors.toMap(Pair::getKey, Pair::getValue)));
+      Pair.of(entry.getValue().getFileGroupId(), Pair.of(entry.getKey(), entry.getValue()))).collect(Collectors.toConcurrentMap(Pair::getKey, Pair::getValue)));
   }
```

Comment:
Do we have concurrent modification to the filesystem view?

Reply:
Yes, multiple partitions can concurrently update the file system view.

Reply:
Not really, Java stream executes the tasks with single thread by default.

Reply:
I have only seen the issue present when using the timeline server. I added a log statement when it is adding the bootstrap base files to the map and the output below shows that the partitions don't seem to be processed sequentially.
```
23/05/22 14:03:58 INFO AbstractTableFileSystemView: Bootstrap Index available for partition partition_path=2015-03-16
23/05/22 14:03:58 INFO AbstractTableFileSystemView: Bootstrap Index available for partition partition_path=2016-03-15
23/05/22 14:03:58 INFO HFileBootstrapIndex: Loaded HFileBasedBootstrapIndex with source base path :/Users/jon/Documents/timelineFail/base
23/05/22 14:03:58 INFO AbstractTableFileSystemView: Bootstrap Index available for partition partition_path=2015-03-17
23/05/22 14:03:58 DEBUG HoodieTableFileSystemView: Adding bootstrap mappping for BaseFile{fullPath=file:/Users/jon/Documents/timelineFail/base/partition_path=2015-03-17/part-00003-8b016653-133e-475c-9820-375c57081c8b.c000.snappy.parquet, fileLen=7673}
23/05/22 14:03:58 DEBUG HoodieTableFileSystemView: Adding bootstrap mappping for BaseFile{fullPath=file:/Users/jon/Documents/timelineFail/base/partition_path=2016-03-15/part-00003-8b016653-133e-475c-9820-375c57081c8b.c000.snappy.parquet, fileLen=7802}
23/05/22 14:03:58 DEBUG HoodieTableFileSystemView: Adding bootstrap mappping for BaseFile{fullPath=file:/Users/jon/Documents/timelineFail/base/partition_path=2015-03-17/part-00002-8b016653-133e-475c-9820-375c57081c8b.c000.snappy.parquet, fileLen=7828}
23/05/22 14:03:58 DEBUG HoodieTableFileSystemView: Adding bootstrap mappping for BaseFile{fullPath=file:/Users/jon/Documents/timelineFail/base/partition_path=2016-03-15/part-00002-8b016653-133e-475c-9820-375c57081c8b.c000.snappy.parquet, fileLen=7686}
23/05/22 14:03:58 DEBUG HoodieTableFileSystemView: Adding bootstrap mappping for BaseFile{fullPath=file:/Users/jon/Documents/timelineFail/base/partition_path=2015-03-17/part-00001-8b016653-133e-475c-9820-375c57081c8b.c000.snappy.parquet, fileLen=7677}
23/05/22 14:03:58 DEBUG HoodieTableFileSystemView: Adding bootstrap mappping for BaseFile{fullPath=file:/Users/jon/Documents/timelineFail/base/partition_path=2015-03-16/part-00003-8b016653-133e-475c-9820-375c57081c8b.c000.snappy.parquet, fileLen=7687}
23/05/22 14:03:58 DEBUG HoodieTableFileSystemView: Adding bootstrap mappping for BaseFile{fullPath=file:/Users/jon/Documents/timelineFail/base/partition_path=2015-03-17/part-00000-8b016653-133e-475c-9820-375c57081c8b.c000.snappy.parquet, fileLen=7690}
23/05/22 14:03:58 DEBUG HoodieTableFileSystemView: Adding bootstrap mappping for BaseFile{fullPath=file:/Users/jon/Documents/timelineFail/base/partition_path=2016-03-15/part-00000-8b016653-133e-475c-9820-375c57081c8b.c000.snappy.parquet, fileLen=7817}
23/05/22 14:03:58 DEBUG HoodieTableFileSystemView: Adding bootstrap mappping for BaseFile{fullPath=file:/Users/jon/Documents/timelineFail/base/partition_path=2015-03-16/part-00002-8b016653-133e-475c-9820-375c57081c8b.c000.snappy.parquet, fileLen=7679}
23/05/22 14:03:58 DEBUG HoodieTableFileSystemView: Adding bootstrap mappping for BaseFile{fullPath=file:/Users/jon/Documents/timelineFail/base/partition_path=2016-03-15/part-00001-8b016653-133e-475c-9820-375c57081c8b.c000.snappy.parquet, fileLen=7684}
23/05/22 14:03:58 DEBUG HoodieTableFileSystemView: Adding bootstrap mappping for BaseFile{fullPath=file:/Users/jon/Documents/timelineFail/base/partition_path=2015-03-16/part-00000-8b016653-133e-475c-9820-375c57081c8b.c000.snappy.parquet, fileLen=7667}
23/05/22 14:03:58 DEBUG HoodieTableFileSystemView: Adding bootstrap mappping for BaseFile{fullPath=file:/Users/jon/Documents/timelineFail/base/partition_path=2015-03-16/part-00001-8b016653-133e-475c-9820-375c57081c8b.c000.snappy.parquet, fileLen=7825}
23/05/22 14:03:58 DEBUG HoodieTableFileSystemView: Adding file-groups for partition :partition_path=2015-03-17, #FileGroups=4
23/05/22 14:03:58 DEBUG AbstractTableFileSystemView: addFilesToView: NumFiles=4, NumFileGroups=4, FileGroupsCreationTime=0, StoreTimeTaken=30
23/05/22 14:03:58 DEBUG AbstractTableFileSystemView: Time to load partition (partition_path=2015-03-17) =161
23/05/22 14:03:58 DEBUG HoodieTableFileSystemView: Adding file-groups for partition :partition_path=2016-03-15, #FileGroups=4
23/05/22 14:03:58 DEBUG HoodieTableFileSystemView: Adding file-groups for partition :partition_path=2015-03-16, #FileGroups=4
23/05/22 14:03:58 DEBUG AbstractTableFileSystemView: addFilesToView: NumFiles=4, NumFileGroups=4, FileGroupsCreationTime=0, 
```

Reply:
Maybe I’m wrong, kind of remember that only the explicit parallel invocation triggers the concurrent execution: https://www.baeldung.com/java-when-to-use-parallel-stream#1-sequential-streams

Reply:
@danny0405 "Java stream executes the tasks with single thread by default." --> This is true.

However, concurrent requests to the timeline server from multiple Spark executors can trigger concurrent execution of `ensurePartitionLoadedCorrectly(partitionPath)`, which can cause the bootstrap index mapping to be concurrently updated.

Reply:
Sorry, still think the fix does not work as expected, the value assign operation like `this.fgIdToPendingCompaction =` is kind of atomic itself, eventhough multiple threads executes the method concurrently, it is still thread safe.

And the entrance of the modification like `fgIdToPendingCompaction` is already guarded by a lock from `#sync` api, more importantly, these internal states of fs view should be kept in compleness like in instant snapshot, we can not put two different states from different instants into one map.

Reply:
The assignment of `fgIdToBootstrapBaseFile` is atomic.  We're not talking about two different timeline instants.  Here for the same instant, the concurrency comes from multiple table partitions updating the same map (`fgIdToBootstrapBaseFile`), and this is observed from the log Jon shows.  That's what this PR fixes.

Reply:
The `fgIdToBootstrapBaseFile` is initialized once in `AbstractTableFileSystemView#addFilesToView` within with we use the Java `stream#map` for updating, the `stream#map` is executed sequencially in single thread, so where does the concurrent updating come from?

Reply:
Synced with @danny0405 offline and resolved the confusion.  `AbstractTableFileSystemView::sync()` syncs the file system view from storage to memory by loading the latest active timeline.  The file system view consistency wrt the latest instant is guarded by the write lock in the reentrant read-write lock in the file system view.  However, the `sync()` does not load file slices or bootstrap base file mapping (`fgIdToBootstrapBaseFile`).  Subsequent partition view calls load file slices against the latest timeline.  Note that, the subsequent partition view calls can be concurrent from multiple Spark executors, which triggers the concurrent execution of `AbstractTableFileSystemView::ensurePartitionLoadedCorrectly(String partition)`.  Inside `ensurePartitionLoadedCorrectly`, `addFilesToView(statuses)` is called to update the bootstrap base file mapping (`fgIdToBootstrapBaseFile`).  Since `fgIdToBootstrapBaseFile` is not a concurrent hash map before, such concurrency leads to updating `fgIdToBootstrapBaseFile` concurrently.  Concurrently updating an ordinary hash map with distinct map keys can still lead to corrupt data in the map.